src/Features/Core/Portable/Diagnostics/EngineV2/CompilerDiagnosticExecutor.cs

```diff
@@ -0,0 +1,292 @@
+
+                                        SetDocument(document);
+                                    }
+                                }
+                                else
+                                {
+                                    // non local diagnostics without location
+                                    _lazyOthers = _lazyOthers ?? new List<DiagnosticData>();
+                                    _lazyOthers.Add(DiagnosticData.Create(_project, diagnostic));
+                                }
+
+                                break;
+                            }
+                        case LocationKind.None:
+                            {
+                                _lazyOthers = _lazyOthers ?? new List<DiagnosticData>();
+                                _lazyOthers.Add(DiagnosticData.Create(_project, diagnostic));
+                                break;
+                            }
+                        case LocationKind.SourceFile:
+                        case LocationKind.MetadataFile:
+                        case LocationKind.XmlFile:
+                            {
+                                // something we don't care
+                                continue;
+                            }
+                        default:
+                            {
+                                Contract.Fail("should not reach");
+                                break;
+                            }
+                    }
+                }
+            }
+
+            public void AddSyntaxDiagnostics(SyntaxTree tree, IEnumerable<Diagnostic> diagnostics)
+            {
+                AddDiagnostics(ref _lazySyntaxLocals, tree, diagnostics);
+            }
+
+            public void AddSemanticDiagnostics(SyntaxTree tree, IEnumerable<Diagnostic> diagnostics)
+            {
+                AddDiagnostics(ref _lazySemanticLocals, tree, diagnostics);
+            }
+
+            public void AddCompilationDiagnostics(IEnumerable<Diagnostic> diagnostics)
+            {
+                Dictionary<DocumentId, List<DiagnosticData>> dummy = null;
+                AddDiagnostics(ref dummy, tree: null, diagnostics: diagnostics);
+
+                // dummy should be always null
+                Contract.Requires(dummy == null);
+            }
+
+            private void AddDiagnostics(
+                ref Dictionary<DocumentId, List<DiagnosticData>> lazyLocals, SyntaxTree tree, IEnumerable<Diagnostic> diagnostics)
+            {
+                foreach (var diagnostic in diagnostics)
+                {
+                    // REVIEW: what is our plan for additional locations? 
+                    switch (diagnostic.Location.Kind)
+                    {
+                        case LocationKind.ExternalFile:
+                            {
+                                // TODO: currently additional file location is not supported.
```

Comment:
yep, we never supported it. we do hold on to additional locations, but doesnt have logic on how to deal with it in engine.